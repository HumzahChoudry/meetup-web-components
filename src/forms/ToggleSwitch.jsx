import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';

import { C_TEXT_PRIMARY, C_TEXT_SECONDARY, C_TEXT_HINT } from 'swarm-constants/dist/js/colorConstants.js';

import Button from './Button';
import Icon from '../media/Icon';
import ToggleWrapper from '../utils/components/ToggleWrapper';

export const TOGGLE_SWITCH_CLASS = 'toggleSwitch';
export const TOGGLE_SWITCH_ACTIVE_CLASS = `${TOGGLE_SWITCH_CLASS}--active`;
export const TOGGLE_SWITCH_DISABLED_CLASS = `${TOGGLE_SWITCH_CLASS}--disabled`;

export const KNOB_CLASS = `${TOGGLE_SWITCH_CLASS}-knob`;
export const KNOB_ACTIVE_CLASS = `${KNOB_CLASS}--active`;
export const KNOB_DISABLED_CLASS = `${KNOB_CLASS}--disabled`;

export const LABEL_CLASS = `${TOGGLE_SWITCH_CLASS}-label`;

/**
 * @module ToggleSwitch
 */
class ToggleSwitch extends React.Component {

	render() {
		const {
			className,
			disabled,
			isActive,
			label,
			labelClassName,
			labelledBy,
			name,
			onCallback,
			offCallback,
			...other
		} = this.props;

		const labelId = labelledBy || `${name}-label`;

		const getActiveIconColor = isActive => isActive ? C_TEXT_PRIMARY : C_TEXT_SECONDARY;
		const getColor = (disbled, isActive) => disabled ? C_TEXT_HINT : getActiveIconColor(isActive);

		return (
			<ToggleWrapper
				isActive={isActive}
				onCallback={onCallback}
				offCallback={offCallback}
			>
				{({
					isActive,
					toggleActive,
				}) => (
					<div {...other}>
						{ label &&
							<label
								className={`${LABEL_CLASS} ${labelClassName} margin--right`}
								id={labelId}
							>
								{label}
							</label>
						}
						<Button
							reset
							disabled={disabled}
							onClick={!disabled && toggleActive}
							className={cx(
								TOGGLE_SWITCH_CLASS,
								{
									[TOGGLE_SWITCH_ACTIVE_CLASS]: isActive,
									[TOGGLE_SWITCH_DISABLED_CLASS]: disabled
								},
								className
							)}
							aria-labelledby={labelId}
							aria-checked={isActive}
							tabIndex={-1}
							role="switch"
						>
							<span className={cx(
								`${KNOB_CLASS} flex flex--center flex--alignCenter`,
								{
									[KNOB_ACTIVE_CLASS]: isActive,
									[KNOB_DISABLED_CLASS]: disabled
								}
							)}>
								<Icon
									shape={isActive ? 'check' : 'cross'}
									color={getColor(disabled, isActive)}
									size='xxs'
									label='Toggle switch label'/>
							</span>
						</Button>
					</div>
				)}
			</ToggleWrapper>
		);
	}
}

ToggleSwitch.propTypes = {
	disabled: PropTypes.bool,
	isActive: PropTypes.bool,
	label: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element
	]),
	labelClassName: PropTypes.string,
	labelledBy: PropTypes.string,
	name: PropTypes.string.isRequired
};

export default ToggleSwitch;
