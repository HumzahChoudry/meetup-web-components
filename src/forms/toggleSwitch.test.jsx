import React from 'react';
import { shallow, mount } from 'enzyme';

import ToggleSwitch, {
	TOGGLE_SWITCH_ACTIVE_CLASS,
	TOGGLE_SWITCH_DISABLED_CLASS,
	TOGGLE_SWITCH_CLASS,
	// KNOB_CLASS,
	KNOB_ACTIVE_CLASS,
	KNOB_DISABLED_CLASS,
	LABEL_CLASS,
} from './ToggleSwitch';

describe('ToggleSwitch', function() {
	const DEFAULT_PROPS = {
		name: 'testToggle'
	};

	const shallowWrapper = (props = {}) =>
		shallow(<ToggleSwitch {...DEFAULT_PROPS} {...props} />);
	const mountWrapper = (props = {}) =>
		mount(<ToggleSwitch {...DEFAULT_PROPS} {...props} />);

	describe('basic', function() {
		it('renders inactive toggle switch', function() {
			expect(shallowWrapper()).toMatchSnapshot();
		});

		it('renders checked toggle switch', function() {
			expect(shallowWrapper({checked: true})).toMatchSnapshot();
		});

		it('has a label only if one is given', () => {
			const component = mountWrapper({label: 'Turn it on?'});
			expect(component.find(`.${LABEL_CLASS}`)).not.toBe(null);
		});

		it('sets `aria-labelledby` attribute if one is given', () => {
			const CUSTOM_LABELLEDBY = 'test-label';
			const component = mountWrapper({labelledBy: CUSTOM_LABELLEDBY});

			expect(component.find(`.${TOGGLE_SWITCH_CLASS}`).props()['aria-labelledby']).toBe(CUSTOM_LABELLEDBY);
		});

		it('adds a class name to a the label element if one is given', () => {
			const CUSTOM_LABEL_CLASS = 'test-label';
			const component = mountWrapper({labelClass: CUSTOM_LABEL_CLASS});
			expect(component.find(`.${CUSTOM_LABEL_CLASS}`)).not.toBe(null);
		});

		it('calls `toggle()` and updates aria-checked when clicked', function() {
			const component = mountWrapper();
			const btnNode = component.find(`.${TOGGLE_SWITCH_CLASS}`);

			expect(btnNode.props()['aria-checked']).toBe(false);
			btnNode.simulate('click');
			expect(btnNode.props()['aria-checked']).toBe(true);
		});
	});

	describe('disabled', function() {
		it(`should add class ${TOGGLE_SWITCH_DISABLED_CLASS} to switch for disabled switches`, function() {
			const component = mountWrapper({disabled: true});
			expect(component.find(`.${TOGGLE_SWITCH_DISABLED_CLASS}`)).not.toBe(null);
		});

		it(`should add class ${KNOB_DISABLED_CLASS} to knob for disabled switches`, function() {
			const component = mountWrapper({disabled: true});
			expect(component.find(`.${TOGGLE_SWITCH_DISABLED_CLASS}`)).not.toBe(null);
		});

		it('does not update aria-checked when clicked when disabled', function() {
			const component = mountWrapper({disabled: true});
			const btnNode = component.find(`.${TOGGLE_SWITCH_CLASS}`);

			expect(btnNode.props()['aria-checked']).toBe(false);
			btnNode.simulate('click');
			expect(btnNode.props()['aria-checked']).toBe(false);
		});
	});

	describe('active', function() {
		it(`should add class ${TOGGLE_SWITCH_ACTIVE_CLASS} to switch for active switches`, function() {
			const component = mountWrapper({disabled: true});
			expect(component.find(`.${TOGGLE_SWITCH_ACTIVE_CLASS}`)).not.toBe(null);
		});

		it(`should add class ${KNOB_ACTIVE_CLASS} to knob for active switches`, function() {
			const component = mountWrapper({disabled: true});
			expect(component.find(`.${TOGGLE_SWITCH_ACTIVE_CLASS}`)).not.toBe(null);
		});
	});

});
