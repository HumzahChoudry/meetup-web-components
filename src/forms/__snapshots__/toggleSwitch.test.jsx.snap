// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleSwitch basic renders checked toggle switch 1`] = `
<span
  aria-pressed={false}
  onKeyUp={[Function]}
  role="button"
  tabIndex="0"
>
  <ToggleSwitch
    checked={true}
    isActive={false}
    name="testToggle"
    toggleActive={[Function]}
  />
</span>
`;

exports[`ToggleSwitch basic renders inactive toggle switch 1`] = `
<span
  aria-pressed={false}
  onKeyUp={[Function]}
  role="button"
  tabIndex="0"
>
  <ToggleSwitch
    isActive={false}
    name="testToggle"
    toggleActive={[Function]}
  />
</span>
`;
